BHoM_Adapter/CRUD/Replace.cs

```diff
@@ -160,40 +188,35 @@ public abstract partial class BHoMAdapter
-            IEnumerable<T> taggedObjects = existingObjects.Where(x => x.Tags.Contains(tag)).ToList(); //ToList() necessary for the method to work correctly. The for each loop below removes the items from the IEnumerable when the tag is removed if not copied to a new list before hand. To be investigated
-            IEnumerable<T> nonTaggedObjects = existingObjects.Where(x => !x.Tags.Contains(tag)).ToList();
-
-            //Remove tag from existing objects
-            foreach (T item in taggedObjects)
-                item.Tags.Remove(tag);
+            IEqualityComparer<T> comparer = Comparer<T>();
+            VennDiagram<T> diagram = Engine.Data.Create.VennDiagram(objsToPush, existingObjs, comparer);
 
-            // Delete object without a tag left
-            Delete(typeof(T), taggedObjects.Where(x => x.Tags.Count == 0).Select(x => x.CustomData[AdapterId]));
+            // Objects to push that do not have any overlap with the existing ones
+            List<T> objsToPush_exclusive = diagram.OnlySet1.ToList();
 
-            // Get objects without the tag that can potentially be merged with the new objects
-            IEqualityComparer<T> comparer = Comparer<T>();
-            VennDiagram<T> diagram1 = Engine.Data.Create.VennDiagram(newObjects, nonTaggedObjects, comparer);
+            // Objects existing in the model that do not have any overlap with the objects being pushed
+            List<T> existingObjs_exclusive = diagram.OnlySet2.ToList();
 
-            // Check and map properties
-            diagram1.Intersection.ForEach(x => x.Item1.MapSpecialProperties(x.Item2, AdapterId));
+            // Discard exclusive existing objects that do not contain the currently specified tag
+            existingObjs_exclusive.RemoveAll(x => !x.Tags.Contains(tag));
 
-            // Get objectsmultiple tags that can potentially be merged with the new objects
-            VennDiagram<T> diagram2 = Engine.Data.Create.VennDiagram(diagram1.OnlySet1, taggedObjects.Where(x => x.Tags.Count > 0), comparer);
+            // Remove the current tag from exclusive existing objects
+            existingObjs_exclusive.ForEach(x => x.Tags.Remove(tag));
 
-            // Check and map properties
-            diagram2.Intersection.ForEach(x => x.Item1.MapSpecialProperties(x.Item2, AdapterId));
+            // Delete exclusive existing objects that do not have any other tag except the current tag from the model
+            Delete(typeof(T), existingObjs_exclusive.Where(x => x.Tags.Count == 0).Select(x => x.CustomData[AdapterId]));
 
-            //Update the tags
-            UpdateProperty(typeof(T), diagram2.OnlySet2.Select(x => x.CustomData[AdapterId]), "Tags", diagram2.OnlySet2.Select(x => x.Tags));
+            //Update the tags for the rest of the existing objects in the model
+            UpdateProperty(typeof(T), existingObjs_exclusive.Select(x => x.CustomData[AdapterId]), "Tags", existingObjs_exclusive.Select(x => x.Tags));
```

Comment:
Think this only should be:

`existingObjs_exclusive.Where(x => x.Tags.Count != 0)`

that need to get the tags updated?

Reply:
That is true! Excellently spotted! Thanks!